sys/vm/vm_map.c

```diff
@@ -6085,6 +6085,63 @@ vm_map_reservation_is_unmapped(vm_map_t map, vm_offset_t reservation)
 	return (false);
 }
 
+/* Return a capability for the reservation that includes base. */
+void * __capability
+vm_map_reservation_cap(vm_map_t map, vm_offset_t base)
+{
+	vm_map_entry_t entry;
+	void * __capability cap;
+	vm_offset_t end, reservation;
+	vm_prot_t max_prot;
+
+	cap = NULL;
+	vm_map_lock_read(map);
+	if (!vm_map_lookup_entry(map, base, &entry))
+		goto out;
+
+	if ((map->flags & MAP_RESERVATIONS) != 0) {
+		reservation = entry->reservation;
+		if (reservation != base) {
```

Comment:
Originally I wasn't planning on doing this check to zoom forward to the start of the reservation so passing the base address (instead of the current address) of the target capability mattered (as I was going to build from the base to the end of the reservation).  However, I relented and instead build a cap for the entire reservation (which also avoids edge cases with representability).  I should probably rename `base` to `va` and change vm_derive_cap to just pass in the address, not the base.